### ISSUE TITLE
PLR1702 spam

### ISSUE THREAD

**Issue by beauxq** [astral-sh/ruff#15570]
If a function has nested blocks to violate PLR1702, almost the entire function is marked for it, including lines that are not violations.

Even lines that are only nested 2 or 3 deep are marked with squigglies.

Even when there are multiple lines in a block nested 9 deep, I think it would be enough to only mark the first line of that block.

![Image](https://github.com/user-attachments/assets/8b03eee8-e446-4218-b73a-d68970e701f4)

**Comment by beauxq** [comment]
PLR1702 is currently marking every line inside this function:
```python
def foo():
    while a:  # should not be marked - not a violation
        if b:  # should not be marked - not a violation
            for c in range(3):  # should not be marked - not a violation
                if d:  # should not be marked - not a violation
                    while e:  # should not be marked - not a violation
                        if f:  # should not be marked - not a violation
                            for g in z:  # not bad to mark this line, but not needed if only marking the max
                                print(p)  # this line should be marked
                                pass  # this line doesn't need to be marked, because the previous line is marked
        else:  # should not be marked - not a violation
            print(q)  # should not be marked - not a violation
```

**Comment by dylwil3** [comment]
That sounds annoying, I'm sorry! When I run this locally or in the playground, I only see a single diagnostic, though the whole range is highlighted. Similarly, when I use the VSCode extension, I see the entire block highlighted but the pop-up only displays a single diagnostic no matter where I hover.

[Playground link](https://play.ruff.rs/1207724a-cad2-45c5-a82e-a6ce79d0cb3b)

VSCode:

<img width="799" alt="Image" src="https://github.com/user-attachments/assets/3883185b-2a9f-4ce2-b5e5-8160eedbd819" />

Terminal:

```console
❯ cat tmp.py
def foo():
    while a:  # should not be marked - not a violation
        if b:  # should not be marked - not a violation
            for c in range(3):  # should not be marked - not a violation
                if d:  # should not be marked - not a violation
                    while e:  # should not be marked - not a violation
                        if f:  # should not be marked - not a violation
                            for g in z:  # not bad to mark this line, but not needed if only marking the max
                                print(p)  # this line should be marked
                                pass  # this line doesn't need to be marked, because the previous line is marked
        else:  # should not be marked - not a violation
            print(q)  # should not be marked - not a violation

❯ uvx ruff check tmp.py --isolated --select PLR --preview --no-cache
tmp.py:2:5: PLR1702 Too many nested blocks (7 > 5)
   |
 1 |   def foo():
 2 | /     while a:  # should not be marked - not a violation
 3 | |         if b:  # should not be marked - not a violation
 4 | |             for c in range(3):  # should not be marked - not a violation
 5 | |                 if d:  # should not be marked - not a violation
 6 | |                     while e:  # should not be marked - not a violation
 7 | |                         if f:  # should not be marked - not a violation
 8 | |                             for g in z:  # not bad to mark this line, but not needed if only marking the max
 9 | |                                 print(p)  # this line should be marked
10 | |                                 pass  # this line doesn't need to be marked, because the previous line is marked
11 | |         else:  # should not be marked - not a violation
12 | |             print(q)  # should not be marked - not a violation
   | |____________________^ PLR1702
   |

Found 1 error.
```

**Comment by beauxq** [comment]
Here's an example with multiple diagnostics.
But I think the multiple diagnostics is not as big of a problem as everything being highlighted.
```python
def foo():
    while a:  # should not be marked - not a violation
        if b:  # should not be marked - not a violation
            for c in range(3):  # should not be marked - not a violation
                if d:  # should not be marked - not a violation
                    while e:  # should not be marked - not a violation
                        if f:  # should not be marked - not a violation
                            for g in z:  # not bad to mark this line, but not needed if only marking the max
                                print(p)  # this line should be marked
                                pass  # this line doesn't need to be marked, because the previous line is marked
                        else:
                            if h:
                                print(r)
        else:  # should not be marked - not a violation
            print(q)  # should not be marked - not a violation
```

**Comment by beauxq** [comment]
In the original example of the first screenshot, there 250 lines all completely highlighted, making it difficult to see other diagnostics.
That's not helpful highlighting.

If there are 5 violations, highlighting 5 lines is enough.

**Comment by dylwil3** [comment]
Thank you, that's helpful!

I think there's a small design question about how to handle this situation, and we should be conscious of whether we are creating a breaking change by modifying where a suppression comment can go. But I agree that we can do better here.

**Comment by MichaReiser** [comment]
I'm torn on this.

Both too-many-branches and complex-structure report the entire function (and I expect so do other too-many-* rules). I also think that highlighting the function itself sort of makes sense, considering:

> Checks for functions or methods with too many nested blocks.

There's a Ruff helper somewhere (I wasn't able to find it just now) that only extracts the range of the function's name. This would reduce the highlighting range but doesn't point the user to the relevant code. 

That's where highlighting the first "violating" block would definitely be an improvement. But this is a breaking change (it's still a preview rule so it might be fine). 

I'm not convinced that highlighting the indent is the right solution. It incorrectly suggests that something is wrong with the indent, which there isn't. 

Side note: The rule should probably be updated to support `match` statements

**Comment by dhruvmanila** [comment]
(This is a scenario where related diagnostics might help where the main diagnostic will highlight the function name but related diagnostics will also highlight the block which makes Ruff consider this as exceeding the limit.)

**Comment by MichaReiser** [comment]
> I do think deep indentation is the problem in this case. Not character-wise, no, but semantically.

Ideally the same approach could be taken for all `too-*` rules and most of them aren't about the indent.
